crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,198 @@
+//! Common types for implementing federation authorization
+
+use headers::{authorization::Credentials, HeaderValue};
+use ruma_common::{ServerName, ServerSigningKeyId};
+use tracing::debug;
+
+/// Typed representation of an `Authorization` header of scheme `X-Matrix`, as defined in the
+/// [Matrix Server-Server API][spec]. Includes an implementation of
+/// [`headers::authorization::Credentials`] for automatically handling the encoding and decoding
+/// when using a web framework that supports typed headers.
+///
+/// [spec]: https://spec.matrix.org/v1.3/server-server-api/#request-authentication
+#[non_exhaustive]
+pub struct XMatrix {
+    /// the server name of the sending server
+    pub origin: Box<ServerName>,
+    /// the server name of the receiving sender. For compatibility with older servers, recipients
+    /// should accept requests without this parameter, but MUST always send it. If this property is
+    /// included, but the value does not match the receiving server’s name, the receiving server
+    /// must deny the request with an HTTP status code 401 Unauthorized.
+    pub destination: Option<Box<ServerName>>,
+    ///the ID, including the algorithm name, of the sending server’s key used to sign the request
+    pub key: Box<ServerSigningKeyId>,
+    /// the signature of the JSON
+    pub sig: String,
```

Comment:
I'm not sure if a plain `String` is what we want here, do we not have a type that can validate these kinds of signatures? Or would this have to wait until `ruma-signatures` is merged into this crate?

Reply:
I think this can be fine-tuned later.